File: Release2/PpFf/tests/StockPrice/StockPrice_PpFf/StockPrice.java

import java.nio.file.Paths;
import java.nio.file.Files;
import java.util.AbstractMap.SimpleEntry;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import java.io.IOException;
import java.util.List;
import static java.util.stream.Collectors.toMap;



/**
 *
 * @author ciubanui
 */
public class StockPrice {
    static final int DEFAULT_NB_ITERATIONS = 5;
    static final String DEFAULT_INPUT_FILE = "../testdata/stock_options_64K.txt";
    
    static public OptionData getOptionData(String data) {
        String structData[] = data.trim().split("\\t|\\s");
        String StockName = structData[0].trim();
        double s = Double.parseDouble(structData[1].trim());
        double strike = Double.parseDouble(structData[2].trim());
        double r = Double.parseDouble(structData[3].trim());
        double divq = Double.parseDouble(structData[4].trim());
        double v = Double.parseDouble(structData[5].trim());
        double t = Double.parseDouble(structData[6].trim());
        int OptionType = (structData[7].trim().charAt(0) == 'P') ? 1 : 0;
        double divs = Double.parseDouble(structData[8].trim());
        double DGrefval = Double.parseDouble(structData[9].trim());
        
        OptionData opt = new OptionData(StockName, s, strike, r, divq, v, t, OptionType, divs, DGrefval);

        return opt;
    }    
    
    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) throws IOException {        
        int nbIterations = DEFAULT_NB_ITERATIONS;
        String inputFile = DEFAULT_INPUT_FILE;

        if (args.length >= 1) {
          inputFile = args[0];
        }

        if (args.length >= 2) {
          nbIterations = Integer.parseInt(args[1]);
        }
        
        List<Map.Entry<String, Double>> stockPrice = null;

	long startTime = System.nanoTime();
       
	for (int i = 0; i < nbIterations; ++i) {
          stockPrice = Files.lines(Paths.get(inputFile))
				.parallel()
                .map(ligne -> StockPrice.getOptionData(ligne))
                .map( opData -> new SimpleEntry<>(opData.StockName, opData.BlkSchlsEqEuroNoDiv()) )
                .collect( toMap(e -> e.getKey(), e -> e.getValue(), (v1, v2) -> v1 > v2 ? v1 : v2) )
                .entrySet().stream()
                .collect( Collectors.toList() );  
	}

     	long duration = (System.nanoTime() - startTime) / nbIterations;
    
    	double milliseconds = (double) duration / 1000000;
    	System.err.println("Temps Java (" + nbIterations + " iterations): " + milliseconds + " ms");       
       
        stockPrice.forEach( x ->
                        System.out.println( x.getKey() + " => " + ( String.format("%.4f", x.getValue() ).replace(",",".") ) ) );
        
    }
    
}

File: Release2/PpFf/tests/StockPrice/StockPrice_PpFf/StockPrice.cpp
#include <iostream>
#include <ff/pipeline.hpp>
#include <ff/farm.hpp>
#include <functional>
#include <regex>
#include <chrono>
#include <ctime>
#include <ratio>
#include <vector>
#include <map>
#include <unordered_map>
#include <string>
#include <sstream>
#include<fstream>
#include <locale>
#include <Pipe.hpp>
#include <unordered_map>
#include <ctype.h>
#include <iomanip>

#include "black_scholes.hpp"

using namespace PpFf;

#define DEFAULT_NB_ITERATIONS 1
#define DEFAULT_INPUT_FILE "/home/iuly/PpFf_JapetServer/tests/StockPrice/testdata/stock_options_64K.txt"
#define DEFAULT_NB_THREADS 1
#define DEFAULT_OUTPUT_FILE_RESULT "/home/iuly/PpFf_JapetServer/tests/StockPrice/testdata/stock_output_result.txt"


struct StockAndPrice {
    std::string StockName;
    double StockPrice;
};

OptionData* getOptionData(std::string* data) {
    std::string name;
    OptionData* opt = new OptionData();
    char otype;
    std::stringstream ins(*data);

    /* read stock option data */
    ins >> opt->StockName;
    ins >> opt->s >> opt->strike >> opt->r >> opt->divq;
    ins >> opt->v >> opt->t >> otype >> opt->divs >> opt->DGrefval;
    opt->OptionType = (otype == 'P');
    
    return opt;
}

StockAndPrice* calculateStockPrice(OptionData* opt) {
    StockAndPrice* stockAndPrice = new StockAndPrice();
    stockAndPrice->StockName = opt->StockName;
    stockAndPrice->StockPrice = BlkSchlsEqEuroNoDiv(opt->s, opt->strike, opt->r, opt->v, 
                                                    opt->t, opt->OptionType, 0);
	
    return stockAndPrice;
}

int main(int argc, char* argv[]) {
    uint32_t nbIterations = DEFAULT_NB_ITERATIONS;
    std::string inputFile = DEFAULT_INPUT_FILE;
	 std::string outputFileResult = DEFAULT_OUTPUT_FILE_RESULT;
    uint32_t nbThreads = DEFAULT_NB_THREADS;

    if (argc >= 2) {
        inputFile = argv[1];
    }

    if (argc >= 3) {
        nbIterations = atoi(argv[2]);
    }
    
    if (argc >= 4) {
        nbThreads = atoi(argv[3]);
    }

    if (argc >= 5) {
        outputFileResult = argv[4];
    }
    
    Reducer<StockAndPrice, double> reducer(0.0, 
                                           [](double maxPrice, StockAndPrice sp) {
                                               return sp.StockPrice > maxPrice ? sp.StockPrice : maxPrice; 
                                           },
                                           [](double max, double workerResult) { 
                                               return workerResult > max ? workerResult : max; 
                                           });
    
    auto begin = std::chrono::high_resolution_clock::now();

    std::unordered_map<std::string, double> currentResult;
    for (uint32_t i = 0; i < nbIterations; ++i) {
        currentResult = 
            Pipe()
            .linesFromFile(inputFile)
            .parallel(nbThreads)
            .map<std::string, OptionData>(getOptionData)
            .map<OptionData, StockAndPrice>(calculateStockPrice)
            //.parallel(1)
            .reduceByKey<StockAndPrice, std::string, double>(reducer, 
                                                             [](StockAndPrice* sp) { return &(sp->StockName); });
        
    }
    auto end = std::chrono::high_resolution_clock::now();
    long duration_ms = std::chrono::duration_cast<std::chrono::milliseconds>(end-begin).count();
    
    fprintf( stderr, "Temps C++  (%3d it.; %2d thr.): %5ld ms {%5ld ms/it. }\n",
             nbIterations, nbThreads, duration_ms, duration_ms / nbIterations );
    
  	 // Write result to file
	 std::ofstream ofs;

  	 ofs.open (outputFileResult, std::ofstream::out | std::ofstream::app);
  	 
	 char buffer[100];
	 snprintf(buffer, sizeof(buffer), "Temps C++  (%3d it.; %2d thr.): %5ld ms {%5ld ms/it. }\n", 
												nbIterations, nbThreads, duration_ms, duration_ms / nbIterations);
	 ofs << buffer;
	 ofs.close();

    std::map<std::string, double> orderedResult;
    for (auto it = currentResult.begin(); it != currentResult.end(); it++) {
    	orderedResult[it->first] = it->second;
    }
    
    for (auto it = orderedResult.begin(); it != orderedResult.end(); it++) {
        std::string currentResultKey = it->first;
        double currentResultValue = it->second;
        std::cout << currentResultKey << " => " << std::fixed << std::setprecision(4) << currentResultValue << std::endl;
    }
   
    return 0;
}
